pkg/model/model.go

```diff
@@ -0,0 +1,279 @@
+}
+
+type Model map[string]*Package
+
+func (m Model) Validate() error {
+	for name, pkg := range m {
+		if name != pkg.Name {
+			return fmt.Errorf("package key %q does not match package name %q", name, pkg.Name)
+		}
+		if err := pkg.Validate(); err != nil {
+			return fmt.Errorf("invalid package %q: %v", pkg.Name, err)
+		}
+	}
+	return nil
+}
+
+type Package struct {
+	Name           string
+	Description    string
+	Icon           *Icon
+	DefaultChannel *Channel
+	Channels       map[string]*Channel
+}
+
+func (m *Package) Validate() error {
+	if m.Name == "" {
+		return errors.New("package name must not be empty")
+	}
+
+	if err := m.Icon.Validate(); err != nil {
+		return fmt.Errorf("invalid icon: %v", err)
+	}
+
+	if len(m.Channels) == 0 {
+		return fmt.Errorf("package must contain at least one channel")
+	}
+
+	if m.DefaultChannel == nil {
+		return fmt.Errorf("default channel must be set")
+	}
+
+	foundDefault := false
+	for name, ch := range m.Channels {
+		if name != ch.Name {
+			return fmt.Errorf("channel key %q does not match channel name %q", name, ch.Name)
+		}
+		if err := ch.Validate(); err != nil {
+			return fmt.Errorf("invalid channel %q: %v", ch.Name, err)
+		}
+		if ch == m.DefaultChannel {
+			foundDefault = true
+		}
+		if ch.Package != m {
+			return fmt.Errorf("channel %q not correctly linked to parent package", ch.Name)
+		}
+	}
+
+	if !foundDefault {
+		return fmt.Errorf("default channel %q not found in channels list", m.DefaultChannel.Name)
+	}
+	return nil
+}
+
+type Icon struct {
+	Data      []byte
+	MediaType string
+}
+
+func (i *Icon) Validate() error {
+	if i == nil {
+		return nil
+	}
+	// TODO(joelanford): Should we check that data and mediatype are set,
+	//   and detect the media type of the data and compare it to the
+	//   mediatype listed in the icon field? Currently, some production
+	//   index databases are failing these tests, so leaving this
+	//   commented out for now.
+	//if len(i.Data) == 0 {
+	//	return errors.New("icon data must be set if icon is defined")
+	//}
+	//if len(i.MediaType) == 0 {
+	//	return errors.New("icon mediatype must be set if icon is defined")
+	//}
+	//return i.validateData()
+	return nil
+}
+
+func (i *Icon) validateData() error {
+	if !filetype.IsImage(i.Data) {
+		return errors.New("icon data is not an image")
+	}
+	t, err := filetype.Match(i.Data)
+	if err != nil {
+		return err
+	}
+	if t.MIME.Value != i.MediaType {
+		return fmt.Errorf("icon media type %q does not match detected media type %q", i.MediaType, t.MIME.Value)
+	}
+	return nil
+}
+
+type Channel struct {
+	Package *Package
+	Name    string
+	Bundles map[string]*Bundle
+}
+
+// TODO(joelanford): This function determines the channel head by finding the bundle that has 0
+//   incoming edges, based on replaces and skips. It also expects to find exactly one such bundle.
+//   Is this the correct algorithm?
+func (c Channel) Head() (*Bundle, error) {
```

Comment:
Answer to your question in the `TODO`. We currently use this ([link](https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/registry/resolver/resolver.go#L591)) to sort bundles in a channel (which effectively finds the head of channel). You should consult with that code and perhaps we should align them.

Reply:
I think that's doing the same thing my code is doing. It's just that my code is not concerned with figuring out the sorted order of the entire channel.

Reply:
Then I think we should be fine with this.

Reply:
@ecordell Did mention that there are index maintainer use cases for potential multi-head channels, so I'll leave this comment in place, but update it to link to the SAT solver implementation and mention Evan's potential use case.